credentials.go

```diff
@@ -0,0 +1,243 @@
+	return endpoints.Endpoint{URI: *uri}, err
+}
+
+func s3Client(ctx context.Context, cfg Config) (*s3.Client, error) {
+	awsAccessKey := "AKIAIOSFODNN7EXAMPLE"
+	awsSecretKey := "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
+	creds := credentials.NewStaticCredentialsProvider(awsAccessKey, awsSecretKey, "")
+	region := config.WithRegion("us-east-2")
+	awsConfig, err := config.LoadDefaultConfig(ctx, region, config.WithCredentialsProvider(creds))
+	if err != nil {
+		return &s3.Client{}, err
+	}
+	client := s3.NewFromConfig(awsConfig, func(o *s3.Options) {
+		o.EndpointResolverV2 = &LocalEndpoint{Bucket: BUCKET, S3Port: cfg.Local.BucketPort}
+	})
+
+	return client, nil
+}
+
+func awsECRClient(ctx context.Context, cfg Config) (*ecr.Client, error) {
+	creds := credentials.NewStaticCredentialsProvider(cfg.Cloud.AWS.AccessKey, cfg.Cloud.AWS.SecretKey, "")
+	region := config.WithRegion(cfg.Cloud.AWS.ClusterRegion)
+	awsConfig, err := config.LoadDefaultConfig(ctx, region, config.WithCredentialsProvider(creds))
+	if err != nil {
+		return &ecr.Client{}, err
+	}
+
+	return ecr.NewFromConfig(awsConfig), nil
+}
+
+// TODO handle expiration date
+func awsECRToken(ctx context.Context, cfg Config) (string, error) {
+	client, err := awsECRClient(ctx, cfg)
+	if err != nil {
+		return "", err
+	}
+	authToken, err := client.GetAuthorizationToken(ctx, &ecr.GetAuthorizationTokenInput{})
+	if err != nil {
+		return "", err
+	}
+
+	if len(authToken.AuthorizationData) < 1 {
+		return "", errors.New("No ECR tokens")
+	}
+
+	data, err := base64.StdEncoding.DecodeString(*authToken.AuthorizationData[0].AuthorizationToken)
+	if err != nil {
+		log.Fatal("error:", err)
+	}
+
+	return string(data), nil
+}
+
+// TODO handle expiration date
+func awsEKSToken(ctx context.Context, cfg Config) (string, error) {
+	creds := credentials.NewStaticCredentialsProvider(cfg.Cloud.AWS.AccessKey, cfg.Cloud.AWS.SecretKey, "")
+	region := config.WithRegion(cfg.Cloud.AWS.ClusterRegion)
+	awsConfig, err := config.LoadDefaultConfig(ctx, region, config.WithCredentialsProvider(creds))
+	if err != nil {
+		return "", err
+	}
+
+	client := sts.NewFromConfig(awsConfig, func(o *sts.Options) {
+		o.APIOptions = []func(*middleware.Stack) error{
+			smithyhttp.AddHeaderValue("x-k8s-aws-id", cfg.Cloud.AWS.ClusterName),
+			smithyhttp.AddHeaderValue("X-Amz-Expires", "60"),
+		}
+	})
+
+	presignedRequest, err := sts.NewPresignClient(client).PresignGetCallerIdentity(ctx, &sts.GetCallerIdentityInput{})
+
+	if err != nil {
+		return "", err
+	}
+
+	return "k8s-aws-v1." + base64.RawURLEncoding.EncodeToString([]byte(presignedRequest.URL)), nil
+}
+
+func kubernetesClient(config Config) (clientcmd.ClientConfig, error) {
+	var client clientcmd.ClientConfig
+	if config.IsLocal {
+		client = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			clientcmd.NewDefaultClientConfigLoadingRules(),
+			&clientcmd.ConfigOverrides{},
+		)
+		return client, nil
+
+	} else {
+		switch config.Cloud.Provider {
```

Comment:
You don't want to create an interface and initialize the proper implementation based on the cloud provider? Supposedly we'll want to support other Kubernetes providers in the future, so we'd have to expand this switch statement.

Reply:
Yes, but this MR is already large enough. I think we should create interfaces in a future MR.